### PR TITLE
Improve database fallbacks

### PR DESCRIPTION
**Pull request by Turbo87**
As suggested on Discord (see https://discord.com/channels/442252698964721669/835156566746595386/923210089085665381), this PR:

- implements a primary database fallback if the replica is unavailable
- renames the existing connection pool methods to `db_read()` and `db_write()`
- implements an additional `db_read_prefer_primary()` method
- uses that new method for endpoints that prefer the primary database, but don't need to write anything to the database

This should improve our resilience a little bit in case of database issues.

### CODE REVIEW

File: src/db.rs
Comment (pietroalbini) [inline src/db.rs:153]: This needs to be updated to mention the primary pool is used even if the replica pool is unavailable.

File: src/db.rs
Comment (pietroalbini) [inline src/db.rs:178]: We can't log a warning every time we failover to the other database pool, otherwise our log traffic will double whenever we have a big outage like yesterday's (eventually dropping logs). I agree we need to have visibility when this happens, but we should do so with a metric we increment every time we need to do a failover.

File: src/db.rs
Comment (pietroalbini) [inline src/db.rs:174]: I'm tempted to say we should failover only when `PoolError::UnhealthyPool` is returned (the amount of connections in the database is 0) instead of any error, otherwise we risk losing visibility into "smaller" bursts of errors.

File: src/db.rs
Comment (pietroalbini) [inline src/db.rs:205]: Same comments as the above method.

**Review by pietroalbini — CHANGES_REQUESTED** [review]

**Comment by Turbo87** [comment]
comments should all be addressed. I've removed the warning for now. we can still add the metrics thing you mentioned later.

**Comment by pietroalbini** [comment]
Could you add a metric as part of this PR? Otherwise we have no visibility when this happens.

File: src/db.rs
Comment (pietroalbini) [inline src/db.rs:186]: Also, just noticed that this returns a read/write connection, not a read-only one. We should make sure connections returned by this method are read-only even when using the primary database, otherwise we might start relying on those being read/write accidentally, causing bugs during failover.

File: src/db.rs
Comment (pietroalbini) [inline src/db.rs:174]: This should also return a read-only connection.

**Review by pietroalbini — CHANGES_REQUESTED** [review]

File: src/db.rs
Comment (Turbo87) [reply src/db.rs:186]: how?

Comment (pietroalbini) [reply src/db.rs:186]: What we could do is store a `ReadOnlyConnection` unit struct in the [connection extensions](https://docs.rs/r2d2/latest/r2d2/struct.PooledConnection.html#method.extensions_mut), indicating whether that connection is currently read-only (unit struct in the extensions) or not. Then, whenever we get a connection from the database we check whether the presence of the struct is correct (present for a RO conn, missing for a RW conn), and otherwise issue the SQL query to switch the connection mode.

Comment (pietroalbini) [reply src/db.rs:186]: Other notes that come through mind:

* We need to make sure this doesn't get into an inconsistent state, so we should also have a `LastSwitchFailed` or something to indicate whether the previous attempt to switch between RO and RW failed, indicating it should be performed unconditionally in the next attempt.
* This could replace the "read-only" section in the pool.
* Do we *need* to cache whether the connection is RW or RO, or can we just send the query to switch the connection mode every time we checkout the connection from the pool? That could be way simpler and less failure prone, but I don't know what the performance impact of it would be off my head.

Comment (Turbo87) [reply src/db.rs:186]: while I generally like the idea of having a type safe way to distinguishing between read-only and writeable database connections, this request is increasing the complexity of this PR by multiple orders of magnitude. I guess ideally this sort of thing would be supported by Diesel itself, instead of us having to monkey patch something like this on top.

I'd appreciate if we could merge this as is for now and then work on further improvements afterwards, instead of blocking this on achieving absolute perfection first. https://en.wikipedia.org/wiki/Perfect_is_the_enemy_of_good 😉

File: src/db.rs
Comment (Turbo87) [reply src/db.rs]: ✅

File: src/db.rs
Comment (pietroalbini) [reply src/db.rs:195]: While in general I agree, for something this crucial to our reliability I'd like to get it right before we merge it.

**Comment by pietroalbini** [comment]
Thinking more about it, this should also have a chaosproxy test to make sure the fallback actually happens as we want.

**Comment by Turbo87** [comment]
feel free to continue this PR, if you want. my goal was to incrementally improve the codebase, but all these additional requirements just to get a small improvement merged make it impossibly hard to contribute.

**Pull request closed without merging**
